Jenkins: Discuss adding a restart job function for use in the OMR Acceptance build

The OMR Acceptance build currently consists of a compile job followed by a sanity job on jdk8 and jdk10 across aix, win, p/z/x linux and xlinux large heap. This equates to 24 downstream jobs, which is 24 points of failure. This number will only grow as we add more platforms.
We often will have one of them fail in the OMR acceptance build for an intermittent infra issue. Ideally we would fix these issues which is why I'm reluctant to suggest adding a feature which allows infra issues to be ignored.

Here is a list to track what has been failing in the recent builds. I will add to this list as I have time and see new failures.

- `#490` - Failed in:
   - AIX 10 compile failure - tracked in #1550
   - Build terminated

- `#489` - Failed in:
   - [Test-Sanity-JDK8-linux_x86-64_cmprssptrs #64](https://ci.eclipse.org/openj9/job/Test-Sanity-JDK8-linux_x86-64_cmprssptrs/64/console)
      - Problem: Initial JenkinsFile clone `Failed to connect to github.com port 443: Connection refused`
      - Solution: unknown
   - [Test-Sanity-JDK8-linux_x86-64 #21](https://ci.eclipse.org/openj9/job/Test-Sanity-JDK8-linux_x86-64/21/console)
      - Problem: full scm checkout on master, which is wrapped in retry(5), failed 5 times. 443 connection refused.
      - Solution: Could try switching to SSH protocol but we would need eclipse to add a key to our bot account
   - Timeout after 8 hours
      - JDK10-linux_x86-64, JDK10-linux_x86-64_cmprssptrs, JDK10-linux_390-64_cmprssptrs didn't complete in time (busy farm)

- `#488` - Failed in:
   - Timeout after 8 hours
      - JDK8-linux_x86-64, JDK10-aix_ppc-64_cmprssptrs, JDK10-linux_x86-64_cmprssptrs, JDK10-linux_x86-64, JDK8-linux_x86-64_cmprssptrs

- `#487` - Failed in:
   - Compile failure JDK10 AIX - tracked in #1550
      - [Build-JDK10-aix_ppc-64_cmprssptrs #117](https://ci.eclipse.org/openj9/job/Build-JDK10-aix_ppc-64_cmprssptrs/117/console)
   - Build terminated

- `#486` - Passed
- `#485` - Passed
- `#484` - Failed in:
   - Windows - No space left on nodes

>existing Jenkins plugin that would allow restarting downstream stages of a pipeline

There are at least 2 open issues I can see with a quick search, both are unresolved. I think the Cloudbees enterprise Jenkins offering has this feature.
- [Restarting stages
](https://issues.jenkins-ci.org/browse/JENKINS-45455)
- [Restart pipeline from specific stage, after failure
](https://issues.jenkins-ci.org/browse/JENKINS-33846)

>How is that reflected in the Jenkins UI?

Your best friend for the restarts will be 
- Slack notifications

You can also find out if a restart is pending on the UI for a build but its not necessarily that obvious.
- The console output will have a link [Input requested]().
This can get buried by other console output.
- From the OMR-Acceptance build number page (the specific build number you are running), there will be a link on the left for [? Paused for input]()
- From the Top level OMR_Acceptance build page, you can see on the "build grid" that a stage is paused. Hovering over will show the restart details.

For the first 2 cases, the link will take you to a page where you can take the restart action. The third option you can action directly from the hover-tip.

>how many of these failures would be resolved by a restart?

Obviously this depends on the failure type. Flaky infra would be the best use case for this.
- Git failures
```
 > git fetch --tags --progress https://github.com/eclipse/openj9.git +refs/pull/*:refs/remotes/origin/pr/*
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from https://github.com/eclipse/openj9.git
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:862)
	at hudson.plugins.git.GitSCM.retrieveChanges(GitSCM.java:1129)
	at hudson.plugins.git.GitSCM.checkout(GitSCM.java:1160)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep.checkout(SCMStep.java:113)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:85)
	at org.jenkinsci.plugins.workflow.steps.scm.SCMStep$StepExecutionImpl.run(SCMStep.java:75)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1$1.call(AbstractSynchronousNonBlockingStepExecution.java:47)
	at hudson.security.ACL.impersonate(ACL.java:290)
	at org.jenkinsci.plugins.workflow.steps.AbstractSynchronousNonBlockingStepExecution$1.run(AbstractSynchronousNonBlockingStepExecution.java:44)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: hudson.plugins.git.GitException: Command "git fetch --tags --progress https://github.com/eclipse/openj9.git +refs/pull/*:refs/remotes/origin/pr/*" returned status code 128:
stdout: 
stderr: remote: Counting objects: 4657, done.        
remote: Compressing objects:   0% (1/161)       
...
...
Resolving deltas: 100% (3402/3402), completed with 1658 local objects.
fatal: unable to look up current user in the passwd file: Bad address
Unexpected end of command stream

	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandIn(CliGitAPIImpl.java:1996)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.launchCommandWithCredentials(CliGitAPIImpl.java:1715)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl.access$300(CliGitAPIImpl.java:72)
	at org.jenkinsci.plugins.gitclient.CliGitAPIImpl$1.execute(CliGitAPIImpl.java:405)
	at hudson.plugins.git.GitSCM.fetchFrom(GitSCM.java:860)
	... 13 more
```
- Space issues
```
java.io.IOException: remote file operation failed: /home/jenkins/workspace/Build-JDK8-linux_x86-64 at hudson.remoting.Channel@582479b6:ub16-x86-3: java.nio.file.FileSystemException: /home/jenkins/workspace/Build-JDK8-linux_x86-64: No space left on device
```
- Random java crashes
```
an exception which occurred:
	in field org.codehaus.groovy.control.MultipleCompilationErrorsException.collector
	in object org.codehaus.groovy.control.MultipleCompilationErrorsException@3b27b666
	in field com.cloudbees.groovy.cps.Outcome.abnormal
	in object com.cloudbees.groovy.cps.Outcome@77d5bc95
	in field org.jenkinsci.plugins.workflow.cps.CpsBodyExecution.outcome
	in object org.jenkinsci.plugins.workflow.cps.CpsBodyExecution@377dec3b
	in field org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution.body
	in object org.jenkinsci.plugins.workflow.steps.TimeoutStepExecution@11e631f1
	in field org.jenkinsci.plugins.workflow.cps.CpsThread.step
	in object org.jenkinsci.plugins.workflow.cps.CpsThread@56ce7ca4
	in field org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.threads
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@250febcb
	in object org.jenkinsci.plugins.workflow.cps.CpsThreadGroup@250febcb
Caused: java.io.NotSerializableException: org.codehaus.groovy.control.ErrorCollector
```
The restarts to watch out for are legitimate intermittent compile or test failures. Restarting blindly can allow bad code to promote and failures to go undocumented.

--- a/buildenv/jenkins/common/pipeline-functions
+++ b/buildenv/jenkins/common/pipeline-functions
@@ -227,10 +227,10 @@ def get_causes_recursively(cause) {
     return cause_string
 }
 
-def build_with_slack(JOB_NAME, PARAMETERS) {
+def build_with_slack(DOWNSTREAM_JOB_NAME, PARAMETERS) {
     def DOWNSTREAM_JOB_NUMBER = ''
     def DOWNSTREAM_JOB_URL = ''
-    def JOB = build job: JOB_NAME, parameters: PARAMETERS, propagate: false
+    def JOB = build job: DOWNSTREAM_JOB_NAME, parameters: PARAMETERS, propagate: false
 
     if (JOB.resultIsWorseOrEqualTo('UNSTABLE')) {
         try {
@@ -248,16 +248,21 @@ def build_with_slack(JOB_NAME, PARAMETERS) {
         build_causes_string = get_causes(currentBuild)
 
         if (JOB.result == "UNSTABLE") {
-            echo "WARNING: Downstream job ${JOB_NAME} is unstable. Job Number: ${DOWNSTREAM_JOB_NUMBER} Job URL: ${DOWNSTREAM_JOB_URL}"
+            echo "WARNING: Downstream job ${DOWNSTREAM_JOB_NAME} is unstable. Job Number: ${DOWNSTREAM_JOB_NUMBER} Job URL: ${DOWNSTREAM_JOB_URL}"
             currentBuild.result = "UNSTABLE"
             if (SLACK_CHANNEL) {
-                slackSend channel: SLACK_CHANNEL, color: 'warning', message: "Unstable: ${JOB_NAME} #${DOWNSTREAM_JOB_NUMBER} (<${DOWNSTREAM_JOB_URL}|Open>)\nStarted by ${env.JOB_NAME} #${env.BUILD_NUMBER} (<${env.BUILD_URL}|Open>)\n${build_causes_string}"
+                slackSend channel: SLACK_CHANNEL, color: 'warning', message: "Unstable: ${DOWNSTREAM_JOB_NAME} #${DOWNSTREAM_JOB_NUMBER} (<${DOWNSTREAM_JOB_URL}|Open>)\nStarted by ${JOB_NAME} #${BUILD_NUMBER} (<${BUILD_URL}|Open>)\n${build_causes_string}"
             }
         } else {
             if (SLACK_CHANNEL) {
-                slackSend channel: SLACK_CHANNEL, color: 'danger', message: "Failure in: ${JOB_NAME} #${DOWNSTREAM_JOB_NUMBER} (<${DOWNSTREAM_JOB_URL}|Open>)\nStarted by ${env.JOB_NAME} #${env.BUILD_NUMBER} (<${env.BUILD_URL}|Open>)\n${build_causes_string}"
+                slackSend channel: SLACK_CHANNEL, color: 'danger', message: "Failure in: ${DOWNSTREAM_JOB_NAME} #${DOWNSTREAM_JOB_NUMBER} (<${DOWNSTREAM_JOB_URL}|Open>)\nStarted by ${JOB_NAME} #${BUILD_NUMBER} (<${BUILD_URL}|Open>)\n${build_causes_string}\nWould you like to restart the job? (<${RUN_DISPLAY_URL}|Open>)"
             }
-            error "Downstream job ${JOB_NAME} did not pass. Job Number: ${DOWNSTREAM_JOB_NUMBER} Job URL: ${DOWNSTREAM_JOB_URL}"
+            timeout(time: RESTART_TIMEOUT, unit: RESTART_TIMEOUT_UNITS) {
+                input message: "Downstream job ${DOWNSTREAM_JOB_NAME} failed. Job Number: ${DOWNSTREAM_JOB_NUMBER} Job URL: ${DOWNSTREAM_JOB_URL}\nRestart failed job '${DOWNSTREAM_JOB_NAME}'?", ok: 'Restart'
+                // If restart is aborted or is timed-out, an error is thrown
+            }
+            // If restart is approved, recursively call this function until we get a pass or a restart-rejection
+            return build_with_slack(DOWNSTREAM_JOB_NAME, PARAMETERS)
         }
     }
     return JOB

--- a/buildenv/jenkins/common/variables-functions
+++ b/buildenv/jenkins/common/variables-functions
@@ -492,6 +492,17 @@ def set_artifactory_config() {
     }
 }
 
+def set_restart_timeout() {
+    RESTART_TIMEOUT = params.RESTART_TIMEOUT
+    if (!RESTART_TIMEOUT) {
+        RESTART_TIMEOUT = VARIABLES.restart_timeout.time
+    }
+    RESTART_TIMEOUT_UNITS = params.RESTART_TIMEOUT_UNITS
+    if (!RESTART_TIMEOUT_UNITS) {
+        RESTART_TIMEOUT_UNITS = VARIABLES.restart_timeout.units
+    }
+}
+
 def set_job_properties(JOB_TYPE) {
     /*************************
     * Setup Jenkins Properties
@@ -629,7 +640,9 @@ def set_job_properties(JOB_TYPE) {
                 'BUILD_NODE':'',
                 'TEST_NODE':'',
                 'PERSONAL_BUILD':'',
-                'SLACK_CHANNEL':''])
+                'SLACK_CHANNEL':'',
+                'RESTART_TIMEOUT':'',
+                'RESTART_TIMEOUT_UNITS':''])
         }
 
         /********************
@@ -713,10 +726,12 @@ def set_job_variables(job_type) {
             set_vendor_variables()
             set_test_targets()
             set_slack_channel()
+            set_restart_timeout()
             break
         case "wrapper":
             //set variable for pipeline all/personal
             set_repos_variables(BUILD_SPECS)
+            set_restart_timeout()
             break
         default:
             error("Unknown Jenkins job type!")

--- a/buildenv/jenkins/jobs/pipelines/Pipeline-Build-Test-All
+++ b/buildenv/jenkins/jobs/pipelines/Pipeline-Build-Test-All
@@ -96,7 +96,7 @@ OPENJDK_SHA = [:]
 BUILD_SPECS = [:]
 builds = [:]
 
-timeout(time: 10, unit: 'HOURS') {
+timeout(time: 12, unit: 'HOURS') {
     timestamps {
         node(LABEL) {
             try {
@@ -221,7 +221,9 @@ def build(OPENJDK_REPO, OPENJDK_BRANCH, SHAS, OPENJ9_REPO, OPENJ9_BRANCH, OMR_RE
                     string(name: 'BUILD_NODE', value: BUILD_NODE),
                     string(name: 'TEST_NODE', value: TEST_NODE),
                     string(name: 'PERSONAL_BUILD', value: PERSONAL_BUILD),
-                    string(name: 'SLACK_CHANNEL', value: SLACK_HANDLE)]
+                    string(name: 'SLACK_CHANNEL', value: SLACK_HANDLE),
+                    string(name: 'RESTART_TIMEOUT', value: RESTART_TIMEOUT),
+                    string(name: 'RESTART_TIMEOUT_UNITS', value: RESTART_TIMEOUT_UNITS)]
         return JOB
     }
 }

--- a/buildenv/jenkins/variables/defaults.yml
+++ b/buildenv/jenkins/variables/defaults.yml
@@ -67,6 +67,9 @@ build_discarder:
     pullRequest: 0
 extra_getsource_options:
   8: '--openssl-version=1.1.1'
+restart_timeout:
+  time: 5
+  units: 'HOURS'
 #========================================#
 # Linux PPCLE 64bits Compressed Pointers
 #========================================#

